test/EngineTest.cpp

```diff
@@ -43,94 +57,259 @@ IdTable makeIdTableFromVector(std::vector<std::array<size_t, TABLE_WIDTH>> table
+  std::array<std::tuple< std::vector<std::array<size_t, 2>>, size_t, std::vector<std::array<size_t, 2>>, size_t, std::vector<std::array<size_t, 3>>> , 3> myTestSet;
+
+  myTestSet[0] = std::make_tuple(
+      std::vector<std::array<size_t, 2>>{{{1, 1}, {1, 3}, {2, 1}, {2, 2}, {4, 1}}},
+      0,
+      std::vector<std::array<size_t, 2>>{{{1, 3}, {1, 8}, {3, 1}, {4, 2}}},
+      0,
+      std::vector<std::array<size_t, 3>>{{1, 1, 3}, {1, 1, 8}, {1, 3, 3}, {1, 3, 8}, {4, 1, 2}}
+  );
+
+  myTestSet[1] = std::make_tuple(
+      std::vector<std::array<size_t, 2>>{{{1, 1}, {1, 3}, {2, 1}, {2, 2}, {4, 1}}},
+      0,
+      std::vector<std::array<size_t, 2>>{{{1, 3}, {1, 8}, {3, 1}, {4, 2}}},
+      0,
+      std::vector<std::array<size_t, 3>>{{1, 1, 3}, {1, 1, 8}, {1, 3, 3}, {1, 3, 8}, {4, 1, 2}, {400000, 200000, 200000}}
+  );
   for (size_t i = 1; i <= 10000; ++i) {
-    a.push_back({I(40000 + i), I(2 + i)});
+    std::get<2>(myTestSet[1]).push_back({4 + i, 2 + i});
   }
-  a.push_back({I(4000001), I(200000)});
-  b.push_back({I(4000001), I(200000)});
-  CALL_FIXED_SIZE_3(lwidth, rwidth, reswidth, J.join, a, 0, b, 0, &res);
-  ASSERT_EQ(2u, res.size());
-
-  a.clear();
-  b.clear();
-  res.clear();
-
-  IdTable c(1, allocator());
-  c.push_back({I(0)});
-
-  ids = {{0, 1}, {0, 2}, {1, 3}, {1, 4}};
-  b = makeIdTableFromVector(ids);
-
-  lwidth = b.cols();
-  rwidth = c.cols();
-  reswidth = b.cols() + c.cols() - 1;
-  // reset the IdTable.
-  res = IdTable(reswidth, allocator());
-  CALL_FIXED_SIZE_3(lwidth, rwidth, reswidth, J.join, b, 0, c, 0, &res);
+  std::get<0>(myTestSet[1]).push_back({400000, 200000});
+  std::get<2>(myTestSet[1]).push_back({400000, 200000});
+
+  myTestSet[2] = std::make_tuple(
+      std::vector<std::array<size_t, 2>>{},
+      0,
+      std::vector<std::array<size_t, 2>>{},
+      0,
+      std::vector<std::array<size_t, 3>>{{40000, 200000, 200000}, {4000001, 200000, 200000}}
+  );
+  for (size_t i = 1; i <= 10000; ++i) {
+    std::get<0>(myTestSet[2]).push_back({4 + i, 2 + i});
+  }
+  std::get<0>(myTestSet[2]).push_back({40000, 200000});
+  std::get<2>(myTestSet[2]).push_back({40000, 200000});
 
-  ASSERT_EQ(2u, res.size());
+  for (size_t i = 1; i <= 10000; ++i) {
+    std::get<0>(myTestSet[2]).push_back({40000 + i, 2 + i});
+  }
+  std::get<0>(myTestSet[2]).push_back({4000001, 200000});
+  std::get<2>(myTestSet[2]).push_back({4000001, 200000});
 
-  ASSERT_EQ(I(0), res(0, 0));
-  ASSERT_EQ(I(1), res(0, 1));
+  return myTestSet;
+}
 
-  ASSERT_EQ(I(0), res(1, 0));
-  ASSERT_EQ(I(2), res(1, 1));
+/* 
+ * @brief The set of join tests for the normal join function had one special
+ * case with different column sizes than the rest, so I had to create another
+ * function just for creating that one.
+ * How to read the formate is described in goThroughSetofTestsWithJoinFunction.
+ */
```

Comment:
If you perform the above refactorings, then it should be rather easy and not a lot of code to also register
more test cases with other combinations of columns etc.